hoot-core/src/main/cpp/hoot/core/ops/GeometryModifierOp.cpp

```diff
@@ -0,0 +1,168 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2015, 2016, 2017, 2018, 2019 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+
+#include "GeometryModifierOp.h"
+#include "hoot-core/src/main/cpp/hoot/core/visitors/geometrymodifiers/GeometryModifierAction.h"
+
+// Hoot
+#include <hoot/core/util/ConfigOptions.h>
+#include <hoot/core/util/Factory.h>
+#include <hoot/core/util/MapProjector.h>
+
+namespace bpt = boost::property_tree;
+using namespace boost;
+
+namespace hoot
+{
+
+HOOT_FACTORY_REGISTER(OsmMapOperation, GeometryModifierOp)
+
+const std::string GeometryModifierAction::FILTER_TAG = "filter";
+const std::string GeometryModifierAction::ARGUMENT_TAG = "arguments";
+
+GeometryModifierOp::GeometryModifierOp()
+{
+  _rulesFileName = ConfigOptions().getGeometryModifierRulesFile();
+
+  // get and instantiate available actions
+  std::vector<std::string> availableActionTypes = Factory::getInstance().getObjectNamesByBase(GeometryModifierAction::className());
+
+  LOG_DEBUG( "Available Geometry Modifiers:")
+  for( std::vector<std::string>::iterator it = availableActionTypes.begin(); it != availableActionTypes.end(); it++ )
+  {
+    shared_ptr<GeometryModifierAction> pAction( Factory::getInstance().constructObject<GeometryModifierAction>(*it) );
+    _actions.append(pAction);
+    LOG_DEBUG( "class: " << *it << " command: " << pAction->getCommandName());
+  }
+}
+
+void GeometryModifierOp::apply(boost::shared_ptr<OsmMap>& map)
+{
+  QList<GeometryModifierActionDesc> actionDescs = _readJsonRules();
+
+  // re-project and set visitor map
+  MapProjector::projectToPlanar(map);
+  _geometryModifierVisitor.setOsmMap(map.get());
+
+  // process
+  foreach (GeometryModifierActionDesc actionDesc, actionDescs)
+  {
+    // visit with specific action, using proper arguments
+    LOG_DEBUG("Processing geometry modifier " + actionDesc.command + "...");
+    _geometryModifierVisitor.setActionDesc(actionDesc);
+    actionDesc.pAction->parseArguments( actionDesc.arguments );
+    map->visitRw(_geometryModifierVisitor);
+  }
+
+  // update operation status info
+  _numAffected = _geometryModifierVisitor._numAffected;
+  _numProcessed = _geometryModifierVisitor._numProcessed;
+}
+
+QList<GeometryModifierActionDesc> GeometryModifierOp::_readJsonRules()
+{
+  // read the json rules
+  bpt::ptree propPtree;
+  bpt::read_json(_rulesFileName.toLatin1().constData(), propPtree );
+
+  QList<GeometryModifierActionDesc> actionDescs;
+
+  BOOST_FOREACH(bpt::ptree::value_type &commandLevelValue, propPtree)
```

Comment:
I don't know if the `ptree` requires `BOOST_FOREACH` or if it works with the C++11 `foreach`, if it does then using `foreach` would be preferred.

Reply:
It works with foreach, replaced.